app/modules/auth/models.py

```diff
@@ -1,49 +1,42 @@
-# encoding: utf-8
-"""
-OAuth2 provider models.
-
-It is based on the code from the example:
-https://github.com/lepture/example-oauth2-server
-
-More details are available here:
-* http://flask-oauthlib.readthedocs.org/en/latest/oauth2.html
-* http://lepture.com/en/2013/create-oauth-server
-"""
-import enum
-
+import time, enum
+from authlib.flask.oauth2.sqla import (
+    OAuth2ClientMixin,
+    OAuth2AuthorizationCodeMixin,
+    OAuth2TokenMixin,
+)
 from sqlalchemy_utils.types import ScalarListType
 
 from app.extensions import db
-from app.modules.users.models import User
 
 
-class OAuth2Client(db.Model):
-    """
-    Model that binds OAuth2 Client ID and Secret to a specific User.
-    """
+class MyOAuth2ClientMixin(OAuth2ClientMixin):
+    def check_requested_scopes(self, scopes):
+        if type(self.scope) == str:
+            allowed = set(self.scope.split())
+        elif type(self.scope) == list:
+            allowed = set(self.scope)
 
-    __tablename__ = 'oauth2_client'
-
-    client_id = db.Column(db.String(length=40), primary_key=True)
-    client_secret = db.Column(db.String(length=55), nullable=False)
+        return allowed.issuperset(set(scopes))
 
-    user_id = db.Column(db.ForeignKey('user.id', ondelete='CASCADE'), index=True, nullable=False)
-    user = db.relationship(User)
+class OAuth2Client(db.Model, MyOAuth2ClientMixin):
+    __tablename__ = 'oauth2_client'
 
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('user.id', ondelete='CASCADE'))
     class ClientTypes(str, enum.Enum):
         public = 'public'
         confidential = 'confidential'
 
     client_type = db.Column(db.Enum(ClientTypes), default=ClientTypes.public, nullable=False)
-    redirect_uris = db.Column(ScalarListType(separator=' '), default=[], nullable=False)
     default_scopes = db.Column(ScalarListType(separator=' '), nullable=False)
```

Comment:
I think you can just drop the `default_scopes`; it seems like it is a simple rename.

Reply:
Well, that's one thing I fixed in my other project. So authlib uses lists differently in the mixins. I'm not a huge fan, but it's convenient to use the mixins. The authlib mixins define scope as a Text and it's a list divided by newlines. Same with grant, and redirect_uri. Then it has @hybrid_property definitions for them that return redirect_uri.splitlines() as an example.

Reply:
I have a little more time now, so I will try to fix based on these comments, and integrate some of the fixes that I made in my own project.